### PR TITLE
feat: Add the ability to prime a breaker with previous stats

### PR DESCRIPTION
**Pull request by lholmquist**
For this first pass, i've only added the ability to import a previous stats object in.  

I'm not sure if we need to import the state of a breaker(closed, open, half-open, etc...) but that is something we can iterate on.

relates to #504

### CODE REVIEW

**Comment by lance** [comment]
Looking good!

> I'm not sure if we need to import the state of a breaker(closed, open, half-open, etc...) but that is something we can iterate on.

If we do this, things get a bit messier. But I think it's probably the right thing to do. Imagine where you're app is under great load, and more and more instances are being scaled up automatically. Now imagine that a circuit in the app has been consistently failing. Each time a new node spins up it will hit the broken endpoint at least once, partially undoing the benefits of the circuit breaker. Obviously, however, the timeout period for entering `HALF_OPEN` would be reset anew with each node that spins up, because I don't think this value can be persisted in a way that lets new instances wait for only as long as the timeout period minus however much time has already elapsed.

I wonder, if instead of `circuit.importStats()` we make it a little more generic. E.g. `circuit.initalize(stats, state)`?

**Comment by lholmquist** [comment]
> I wonder, if instead of circuit.importStats() we make it a little more generic. E.g. circuit.initalize(stats, state

I'm good with that.

And I agree that we should probably be initializing breakers to the state in which they where when the function/pod/etc..  was destroyed.  I might make that a separate PR so this one can lay the groundwork and not get to bloated

**Comment by lholmquist** [comment]
I update the `initialize` function signature to be an object,  so it will be easier to pass in both the stats and the state in the future

**Comment by lance** [comment]
> I update the `initialize` function signature to be an object, so it will be easier to pass in both the stats and the state in the future

Why not just accept a `Status` object with `status` as an option in the constructor?

```js
const Circuit = require('opossum')

// Because we only export the CircuitBreaker class we have to do this kind of clunky stuff
const status = Circuit.newStatus({ stats })
const circuit = new Circuit(myfn, { status })
```

This would allow us to do some sanity checks when the Status instance is initalized. What do you think? Too clunky?

**Comment by lholmquist** [comment]
That first example doesn't look to clunky and i would like to not have to do a Major bump if we don't have to

Let me refactor what i did to see how it looks

I would rather save that for when we convert to a pure ESM module :)

**Review by helio-frota — APPROVED** [review]
LGTM.

**Comment by lholmquist** [comment]
@lance just pushed a refactor.

now it looks like this:

```
const statusOptions = {}
statusOptions.stats = {.....}
const newStatus = CircuitBreaker.newStatus(statusOptions);

const breaker = new CircuitBreaker(fn, {status: newStatus})

```


I still need to update the readme,  but how is this looking.  I like this better than having that initialization function.

**Comment by lholmquist** [comment]
@lance @helio-frota  i think i'm ready for a review here with this new refactor.

File: README.md [hunk 1]
```diff
@@ -87,6 +87,56 @@ const breaker = new CircuitBreaker(delay);
...
 breaker.fallback((delay, a, b, c) => `Sorry, out of service right now. But your parameters are: ${delay}, ${a}, ${b} and ${c}`);
 ```
+### Status Initialization
+
+There may be times where you will need to pre-populate the stats of the Circuit Breaker Status Object.  The major use case for this is in a serverless functions environment,  or in a container based deployment, where the container being deployed is ephemeral.
+
+Getting the existing cumalative stats for a breaker can be done like this:
+
+```
+const stats = breaker.status.stats;
```
Comment (lance) [inline README.md:97]:
```suggestion
const stats = breaker.stats;
```

There's actually a convenience function for this on `Circuit`

File: README.md [hunk 2]
```diff
@@ -87,6 +87,56 @@ const breaker = new CircuitBreaker(delay);
 breaker.fire(20000, 1, 2, 3);
 breaker.fallback((delay, a, b, c) => `Sorry, out of service right now. But your parameters are: ${delay}, ${a}, ${b} and ${c}`);
 ```
+### Status Initialization
+
+There may be times where you will need to pre-populate the stats of the Circuit Breaker Status Object.  The major use case for this is in a serverless functions environment,  or in a container based deployment, where the container being deployed is ephemeral.
```
Comment (lance) [inline README.md:92]:
```suggestion
There may be times where you will need to pre-populate the stats of the Circuit Breaker Status Object.  Primary use cases for this are in a serverless environment such as Knative or AWS Lambda, or any container based platform, where the container being deployed is ephemeral.
```
A couple of wording tweaks here, to emphasize that this isn't just for functions.

**Review by lance — COMMENTED** [review]
Just a few wording suggestions

File: lib/circuit.js [hunk 3]
```diff
@@ -138,7 +154,16 @@ class CircuitBreaker extends EventEmitter {
     this[VOLUME_THRESHOLD] = Number.isInteger(options.volumeThreshold)
       ? options.volumeThreshold : 0;
     this[WARMING_UP] = options.allowWarmUp === true;
-    this[STATUS] = new Status(this.options);
+
+    // The user can pass in a Status object ot inialize the Status/stats
+    if (this.options.status && this.options.status instanceof Status) {
+      // Do a check that this is a Status Object,
+      // should it error if it is not a Status Object
```
Comment (lance) [inline lib/circuit.js:161]: I'm leaning towards yes. It seems wrong to just accept bad input and not do anything about it.

File: README.md (hunk 1 above)
Comment (lholmquist) [reply README.md]: i knew that :)

File: lib/circuit.js (hunk 3 above)
Comment (lholmquist) [reply lib/circuit.js:161]: same,  i'll update to throw an error if the status option is not an instance of the Status class

Comment (lholmquist) [reply lib/circuit.js]: updated

File: test/status-test.js [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    latencyTimes: []
+  };
+
+  const status = CircuitBreaker.newStatus({stats: prevStats});
+  t.equal(status instanceof Status, true, 'returns a new Status instance');
+
+  const stats = status.stats;
+  t.equal(stats.failures, 1, 'status reports 1 failure');
+  t.equal(stats.rejects, 1, 'status reports 1 reject');
+  t.equal(stats.fires, 1, 'status reports 5 fires');
```
Comment (helio-frota) [inline test/status-test.js:31]: status reports 5 fires, should be 1 fire?

**Review by helio-frota — CHANGES_REQUESTED** [review]

File: test/status-test.js (hunk 4 above)
Comment (lholmquist) [reply test/status-test.js:31]: yea,  i can't copy/paste very good

Comment (lholmquist) [reply test/status-test.js]: Updated

**Review by helio-frota — APPROVED** [review]

**Review by lance — APPROVED** [review]
Thanks for tackling this @lholmquist - looks good. :)

**Pull request merged**